### PR TITLE
Refactor inter-tab notifications

### PR DESCRIPTION
**Pull request by bfollington**
Fixes https://github.com/subconsciousnetwork/subconscious/issues/937
FIxes https://github.com/subconsciousnetwork/subconscious/issues/978

All core operations (save, delete, move, merge, change audience) are now handled at the app level. Errors are handled at the root and will pop toast notifications.

Child views pass a request for the operation up the tree, the app store runs the effect and then passes the success notification back down the tree. 

In playtesting this seems to resolve all our inter-tab desync issues.

There's a lot of churn here so I was perhaps a little overzealous with the testing 😅 

---

- [x] Save notification
- [x] Refresh Deck
- [x] Merge notification
- [x] Move notification
- [x] Change audience notification
- [x] Flow request actions all the way up (do not stop at DetailStack)
    - [x] Deck
    - [x] Notebook
    - [x] Profile
- [x] Figure out why metasheet isn't updating
- [x] Fix tests
- [x] Test mapping
   - [x] Cursors (detailstack <-> tab root)
   - [x] AppAction from and to

### CODE REVIEW

File: xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift [hunk 1]
```diff
@@ -281,29 +285,54 @@ enum MemoEditorDetailNotification: Hashable {
...
+//        case let .updateAudience(address, audience):
+//            return .updateAudience(address: address, audience: audience)
+//        default:
+//            return nil
+//        }
+//    }
+}
+
 extension MemoEditorDetailNotification {
     static func from(_ action: MemoEditorDetailAction) -> Self? {
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift:324]: add test?

File: xcode/Subconscious/Shared/Components/AppView.swift [hunk 2]
```diff
@@ -1206,6 +1217,90 @@ struct AppModel: ModelProtocol {
...
+                state: state,
+                environment: environment,
+                address: address,
+                audience: audience
+            )
+        case .failSaveEntry(address: let address, error: let error):
+            logger.warning("""
+                Failed to save entry: \(address)
+                \(error)
+                """)
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/AppView.swift:1259]: Nit: indendation. Swift multiline strings are sensitive to the indentation of opening and closing quotes. They need to match text indentation. Should be:

```swift
logger.warning(
    """
    Failed to save entry: \(address)
    \(error)
    """
)
```

File: xcode/Subconscious/Shared/Components/AppView.swift [hunk 3]
```diff
@@ -1206,6 +1217,90 @@ struct AppModel: ModelProtocol {
...
+                action: .pushToast(
+                    message: "Could not save note"
+                ),
+                environment: environment
+            )
+        case .failMoveEntry(from: let from, to: let to, error: let error):
+            logger.warning("""
+                Failed to move entry: \(from) -> \(to)
+                \(error)
+                """)
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/AppView.swift:1272]: Nit: indentation. See above.

File: xcode/Subconscious/Shared/Components/AppView.swift [hunk 4]
```diff
@@ -1206,6 +1217,90 @@ struct AppModel: ModelProtocol {
...
+                action: .pushToast(
+                    message: "Could not move note"
+                ),
+                environment: environment
+            )
+        case .failMergeEntry(parent: let parent, child: let child, error: let error):
+            logger.warning("""
+                Failed to merge entries: \(child) -> \(parent)
+                \(error)
+                """)
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/AppView.swift:1284]: Nit: indentation. See above.

File: xcode/Subconscious/Shared/Components/AppView.swift [hunk 5]
```diff
@@ -1206,6 +1217,90 @@ struct AppModel: ModelProtocol {
...
+                action: .pushToast(
+                    message: "Could not merge notes"
+                ),
+                environment: environment
+            )
+        case .failUpdateAudience(address: let address, audience: let audience, error: let error):
+            logger.warning("""
+                Failed to update audience for entry: \(address) \(audience)
+                \(error)
+                """)
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/AppView.swift:1296]: Nit: indentation. See above.

File: xcode/Subconscious/Shared/Components/AppView.swift [hunk 6]
```diff
@@ -1206,6 +1217,90 @@ struct AppModel: ModelProtocol {
...
+            return update(
+                state: state,
+                action: .pushToast(
+                    message: "Could not merge notes"
+                ),
+                environment: environment
+            )
+        case .failUpdateAudience(address: let address, audience: let audience, error: let error):
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/AppView.swift:1292]: Nit: shorten this to `case let .failUpdateAudience(address, audience, error):`

File: xcode/Subconscious/Shared/Components/AppView.swift [hunk 7]
```diff
@@ -2686,7 +2781,7 @@ struct AppModel: ModelProtocol {
         let fx: Fx<Action> = environment.data
             .deleteMemoPublisher(address)
             .map({ _ in
-                Action.succeedDeleteMemo(address)
+                Action.succeedDeleteEntry(address)
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/AppView.swift:2784]: 👍 thanks for making this name change for consistency

File: xcode/Subconscious/Shared/Components/Deck/DeckView.swift [hunk 8]
```diff
@@ -274,6 +324,67 @@ struct DeckModel: ModelProtocol {
                 state: state,
                 environment: environment
             )
+        case .refreshUpcomingCards:
+            return refreshEachCard(state: state, environment: environment)
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Deck/DeckView.swift:328]: Nit: rename function to match action?

File: xcode/Subconscious/Shared/Components/Deck/DeckView.swift [hunk 9]
```diff
@@ -338,6 +449,49 @@ struct DeckModel: ModelProtocol {
...
+                    }
+                    
+                    updated.append(card.update(entry: entry))
+                }
+
+                // Set the combined deck and return
+                return .setDeck(updated, state.pointer)
+            }
+            .recover({ error in
+                .noCardsToDraw
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Deck/DeckView.swift:488]: This action name is delightfully prose-y 👍

File: xcode/Subconscious/Shared/Components/Deck/DeckView.swift [hunk 10]
```diff
@@ -361,23 +515,33 @@ struct DeckModel: ModelProtocol {
...
+            state: Self,
+            environment: Environment,
+            deck: [CardModel],
+            startIndex: Int
+        ) -> Update<Self> {
             var model = state
             model.deck = deck
             model.seen = Set(deck.compactMap { card in card.entry })
-            model.pointer = 0
+            model.pointer = startIndex
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Deck/DeckView.swift:527]: Stray thought: do we do bounds-checking on this index?

File: xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift [hunk 11]
```diff
@@ -136,18 +136,19 @@ enum DetailStackAction: Hashable {
...
-    /// Gets forwarded up to parent for handling.
-    case requestDeleteMemo(Slashlink?)
-    /// Deletion attempt failed. Forwarded down from parent.
-    case failDeleteMemo(String)
-    /// Deletion attempt succeeded. Forwarded down from parent.
-    case succeedDeleteMemo(Slashlink)
-
+    /// Note lifecycle events.
+    /// `request`s are passed up to the app root
+    /// `succeed`s are passed down from the app root
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift:141]: Thanks for the comment 👍

File: xcode/Subconscious/Shared/Library/DummyDataUtilities.swift [hunk 12]
```diff
@@ -184,6 +184,21 @@ extension EntryStub: DummyData {
...
+            contentType: "text/subtext",
+            created: Date.now,
+            modified: Date.now,
+            fileExtension: "subtext",
+            additionalHeaders: Headers(),
+            body: String.dummyDataMedium()
+        )
+    }
+    
+
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Library/DummyDataUtilities.swift:199]: Minor nit: extra space

File: xcode/Subconscious/Shared/UIKit/BlockEditor/BlockEditorModel.swift [hunk 13]
```diff
@@ -137,12 +137,9 @@ extension BlockEditor {
         /// have in cache.
         case refreshTranscludesFor(id: UUID)
         /// Save a snapshot
-        case save(_ snapshot: MemoEntry?)
+        case requestSave(_ snapshot: MemoEntry?)
+        case forwardRequestSave(_ snapshot: MemoEntry)
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/UIKit/BlockEditor/BlockEditorModel.swift:141]: What is forwardRequestSave?

**Review by gordonbrander — APPROVED** [review]
LGTM. A few nits and questions

File: xcode/Subconscious/Shared/UIKit/BlockEditor/BlockEditorModel.swift (hunk 13 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/UIKit/BlockEditor/BlockEditorModel.swift:141]: Since `requestSave` can be emitted as a flow-on from other actions it might not actually be emitted on the `actions` stream of `Store<BlockEditorModel>`. `forwardRequestSave` is an explicitly emitted action, as an effect, to ensure this request is always forwarded up to the app-level.

File: xcode/Subconscious/Shared/Components/Deck/DeckView.swift (hunk 8 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Deck/DeckView.swift:328]: Good catch!
